runtime/descriptor.go

```diff
@@ -98,24 +98,62 @@ func propertySet(f *Frame, desc, inst, value *Object) *BaseException {
 
 // makeStructFieldDescriptor creates a descriptor with a getter that returns
 // the field given by fieldName from t's basis structure.
-func makeStructFieldDescriptor(t *Type, fieldName, propertyName string) *Object {
+func makeStructFieldDescriptor(t *Type, fieldName, propertyName string, writtable bool) *Object {
 	field, ok := t.basis.FieldByName(fieldName)
 	if !ok {
 		logFatal(fmt.Sprintf("no such field %q for basis %s", fieldName, nativeTypeName(t.basis)))
 	}
+
 	getterFunc := func(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
 		var ret *Object
 		var raised *BaseException
-		if raised = checkFunctionArgs(f, fieldName, args, ObjectType); raised == nil {
-			o := args[0]
-			if !o.isInstance(t) {
-				format := "descriptor '%s' for '%s' objects doesn't apply to '%s' objects"
-				raised = f.RaiseType(TypeErrorType, fmt.Sprintf(format, propertyName, t.Name(), o.typ.Name()))
-			} else {
-				ret, raised = WrapNative(f, t.slots.Basis.Fn(o).FieldByIndex(field.Index))
-			}
+
+		if raised = checkFunctionArgs(f, fieldName, args, ObjectType); raised != nil {
+			return ret, raised
 		}
+
+		self := args[0]
+		if !self.isInstance(t) {
+			format := "descriptor '%s' for '%s' objects doesn't apply to '%s' objects"
+			raised = f.RaiseType(TypeErrorType, fmt.Sprintf(format, propertyName, t.Name(), self.typ.Name()))
+			return ret, raised
+		}
+
+		ret, raised = WrapNative(f, t.slots.Basis.Fn(self).FieldByIndex(field.Index))
 		return ret, raised
 	}
-	return newProperty(newBuiltinFunction("_get"+fieldName, getterFunc).ToObject(), None, None).ToObject()
+
+	setterFunc := func(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+		ret := None
+		var raised *BaseException
+		if raised = checkFunctionArgs(f, fieldName, args, ObjectType, ObjectType); raised != nil {
+			return ret, raised
+		}
+
+		self := args[0]
+		newvalue := args[1]
+		if !newvalue.isInstance(getNativeType(field.Type)) {
+			format := "descriptor '%s' for '%s' objects doesn't apply to '%s' objects"
+			raised = f.RaiseType(TypeErrorType, fmt.Sprintf(format, propertyName, t.Name(), newvalue.typ.Name()))
+			return ret, raised
+		}
+
+		var converted reflect.Value
+		val := t.slots.Basis.Fn(self).FieldByIndex(field.Index)
+
+		if converted, raised = maybeConvertValue(f, newvalue, val.Type()); raised != nil {
+			return ret, raised
+		}
+
+		val.Set(converted)
+		return ret, raised
+	}
+
+	getter := newBuiltinFunction("_get"+fieldName, getterFunc).ToObject()
+	setter := None
+	if writtable {
+		setter = newBuiltinFunction("_set"+fieldName, setterFunc).ToObject()
```

Comment:
In that case I was trying to avoid multiple returns that were very similar:

```
  return newProperty(...)
}
return newProperty(...)
```

Where `...` had a lot of repetition. I think what you have now is an improvement, I'm just suggesting moving the setterFunc definition inside the if block.